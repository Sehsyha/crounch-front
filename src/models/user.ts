export interface SignupResponse {
  id: string;
  email: string;
}

export interface LoginResponse {
  accessToken: string;
}
