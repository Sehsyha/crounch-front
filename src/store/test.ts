import { ActionTree, MutationTree, GetterTree, Module, ActionContext } from 'vuex';
import { RootState } from '.';

export type ActionCall<S, R> = (injectee: ActionContext<S, R>, payload?: any) => any;

export function createStoreModuleMock<S, A, M, G>(initialState: S, actionKeys: A, mutationKeys: M, getterKeys: G): Module<S, RootState> {
  const actions: ActionTree<S, RootState> = {};
  const mutations: MutationTree<S> = {};
  const getters: GetterTree<S, RootState> = {};

  Object.keys(actionKeys).forEach(key => actions[key] = jest.fn());
  Object.keys(mutationKeys).forEach(key => mutations[key] = jest.fn());
  Object.keys(getterKeys).forEach(key => getters[key] = jest.fn());

  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: initialState,
  }
}
