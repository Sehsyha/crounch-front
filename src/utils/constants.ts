export const TOKEN_STORAGE_KEY = 'token';

export const DEFAULT_CATEGORY_ID = '00000000-0000-0000-0000-000000000000'
export const DEFAULT_CATEGORY_NAME = 'Divers'
