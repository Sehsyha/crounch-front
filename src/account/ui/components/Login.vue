<template>
  <div id="login">
    <h2>Connexion</h2>
    <form @submit.prevent="login">
      <div v-if="!isEmailValid" class="error" id="email-error">
        Please enter a valid email.
      </div>
      <div v-if="!isPasswordValid" class="error" id="password-error">
        Please enter a password.
      </div>
      <div v-if="loginSuccess" class="success">
        You've logged in successfully !!
      </div>
      <input type="text" placeholder="Email" v-model="email" />
      <input type="password" placeholder="Password" v-model="password" />
      <input type="submit" value="Se connecter" />
    </form>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { getModule } from 'vuex-module-decorators';

import { validateEmail } from '@/utils/form-validation';
import { AccountModule } from '@/account/store/AccountModule';

@Component
export default class Login extends Vue {
  public accountModule: AccountModule = getModule(AccountModule)

  email: string = '';
  password: string = '';
  loginSuccess: boolean = false;

  login() {
    if (!this.isEmailValid || !this.isPasswordValid) {
      return;
    }

    const { email, password } = this;

    this.accountModule.loginAction({ email, password }).then(() => {
      this.$router.replace('/lists');
    });
  }

  get isEmailValid(): boolean {
    return !!validateEmail(this.email);
  }

  get isPasswordValid(): boolean {
    return this.password.length !== 0;
  }
}
</script>
