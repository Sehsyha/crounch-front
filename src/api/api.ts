import { getAPIURL } from '@/utils/environment';
import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { FetchError } from '@/utils/error';
import { AuthModule } from '@/store/AuthModule';
import { getModule } from 'vuex-module-decorators';
import router from '@/router/router';

export interface FetchOptions {
  url: string;
  data?: object;
  method: string;
}

export function doFetch<T>(options: FetchOptions): Promise<T> {
  return fetch(getUrl(options.url), getFetchOptions(options))
    .then(res => {
      if (res.status >= 200 && res.status <= 299 && res.status !== 204) {
        return res.json()
      }
      if (res.status === 204) {
        return
      }
      if (res.status === 401) {
        const authModule = getModule(AuthModule)
        authModule.logoutAction().then(() => {
          router.push({ name: 'home' })
        })
      }

      return res.json()
        .then(body => { throw new FetchError('Erreur lors de la requÃªte', res.status, body) })
        .catch()
    })
}

function getFetchOptions({ data, method }: FetchOptions): object {
  let options: any = {
    method,
    headers: {
      Authorization: localStorage.getItem(TOKEN_STORAGE_KEY),
    },
  };

  if (method !== 'GET') {
    options.body = JSON.stringify(data);
  }

  return options;
}

function getUrl(url: string): string {
  return `${getAPIURL()}/${url}`;
}
