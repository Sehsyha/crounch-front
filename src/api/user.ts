import { api } from './api';

export interface SignupResponse {
  id: string;
  email: string;
}

export interface LoginResponse {
  accessToken: string;
}

export function signup(email: string, password: string): Promise<SignupResponse> {
  return api
    .post<SignupResponse>('users', {
      email,
      password,
    })
    .then(res => res.data);
}

export function login(email: string, password: string): Promise<LoginResponse> {
  return api
    .post<LoginResponse>('users/login', {
      email,
      password,
    })
    .then(res => res.data);
}
