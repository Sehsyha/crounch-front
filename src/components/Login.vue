<template>
  <div id="login">
    <h2>Connexion</h2>
    <form @submit.prevent="login">
      <div v-if="!isEmailValid" class="error" id="email-error">
        Please enter a valid email.
      </div>
      <div v-if="!isPasswordValid" class="error" id="password-error">
        Please enter a password.
      </div>
      <div v-if="loginSuccess" class="success">
        You've logged in successfully !!
      </div>
      <input type="text" placeholder="Email" v-model="email" />
      <input type="password" placeholder="Password" v-model="password" />
      <input type="submit" value="Se connecter" />
    </form>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { getModule } from 'vuex-module-decorators';

import { login } from '@/api/user';
import { validateEmail } from '@/utils/form-validation';
import { AuthModule } from '@/store/auth/AuthModule';

@Component
export default class Login extends Vue {
  public authModule: AuthModule = getModule(AuthModule)

  email: string = '';
  password: string = '';
  loginSuccess: boolean = false;

  login() {
    if (!this.isEmailValid || !this.isPasswordValid) {
      return;
    }

    const { email, password } = this;

    this.authModule.loginAction({ email, password }).then(() => {
      this.$router.replace('/lists');
    });
  }

  get isEmailValid(): boolean {
    return !!validateEmail(this.email);
  }

  get isPasswordValid(): boolean {
    return this.password.length !== 0;
  }
}
</script>
