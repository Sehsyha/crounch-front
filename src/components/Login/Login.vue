<template>
  <div id="login">
    <h2>Login</h2>
    <form @submit.prevent="login">
      <div v-if="!isEmailValid" class="error" id="email-error">Please enter a valid email.</div>
      <div v-if="!isPasswordValid" class="error" id="password-error">Please enter a password.</div>
      <div v-if="loginSuccess" class="success">You've logged in successfully !!</div>
      <input type="text" placeholder="Email" v-model="email" />
      <input type="password" placeholder="Password" v-model="password" />
      <input type="submit" value="Se connecter" />
    </form>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { Action } from 'vuex-class';

import { login } from '@/api/user';
import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { validateEmail } from '@/utils/form-validation';
import { AuthActions } from '@/store/auth/keys';
import { authNamespace } from '@/store/auth';

@Component
export default class Login extends Vue {
  @Action(AuthActions.LOGIN, authNamespace) doLogin: any;
  email: string = '';
  password: string = '';
  loginSuccess: boolean = false;

  login() {
    if (!this.isEmailValid || !this.isPasswordValid) {
      return;
    }
    const { email, password } = this;

    this.doLogin({ email, password }).then(() => {
      this.$router.push('/lists');
    });
  }

  get isEmailValid(): boolean {
    return !!validateEmail(this.email);
  }

  get isPasswordValid(): boolean {
    return this.password.length !== 0;
  }
}
</script>
