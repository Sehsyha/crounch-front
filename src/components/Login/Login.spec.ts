import { Wrapper } from '@vue/test-utils';
import { when } from 'jest-when';

import { login } from '@/api/user';
import Login from '@/components/Login/Login.vue';
import { validateEmail } from '@/utils/form-validation';
import { shallowComponent } from '@/utils/test';
import { TOKEN_STORAGE_KEY } from '@/utils/constants';

jest.mock('@/api/user');
jest.mock('@/utils/form-validation');

describe('Login', () => {
  let wrapper: Wrapper<Login>;
  let setItemMock: jest.SpyInstance;
  const email = 'a';
  const password = 'password';
  const token = 'auth-token';

  beforeEach(() => {
    wrapper = shallowComponent(Login);
    (login as jest.Mock).mockResolvedValue({ accessToken: token });
    setItemMock = jest.spyOn(Storage.prototype, 'setItem');
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('Init', () => {
    it('Should initialize email to empty string.', () => {
      expect(wrapper.vm.$data.email).toBe('');
    });

    it('Should initialize password to empty string.', () => {
      expect(wrapper.vm.$data.password).toBe('');
    });

    it('Should initialize loginSuccess to false.', () => {
      expect(wrapper.vm.$data.loginSuccess).toBe(false);
    });
  });

  describe('Form validation', () => {
    describe('Email', () => {
      it('Should render an error when email is not valid.', () => {
        when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(false);
        wrapper.setData({ email });
        expect(wrapper.find('#email-error').exists()).toBeTruthy();
      });

      it('Should not render an error when email is valid.', () => {
        when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);
        wrapper.setData({ email });
        expect(wrapper.find('#email-error').exists()).toBeFalsy();
      });
    });

    describe('Password', () => {
      it('Should render an error when password input is not filled.', () => {
        expect(wrapper.find('#password-error').exists()).toBeTruthy();
      });

      it('Should not render an error when password input is filled.', () => {
        wrapper.setData({ password: 'az' });

        expect(wrapper.find('#password-error').exists()).toBeFalsy();
      });
    });
  });

  describe('Form submission', () => {
    it('Should not call api when email is in error and not password.', () => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(false);
      wrapper.setData({ password: 'password', email });

      wrapper.find('input[type=submit]').trigger('click');

      expect(login as jest.Mock).not.toHaveBeenCalled();
    });

    it('Should not call api when password is in error and not email.', () => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);
      wrapper.setData({ password: '', email });

      wrapper.find('input[type=submit]').trigger('click');

      expect(login as jest.Mock).not.toHaveBeenCalled();
    });

    it('Should call api with right parameters when everything is valid.', () => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);

      wrapper.setData({ password, email });

      wrapper.find('[type=submit]').trigger('click');

      expect(login as jest.Mock).toHaveBeenCalledWith(email, password);
    });

    it('Should display success message when user has signed up.', done => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);

      wrapper.setData({ password, email });

      wrapper.find('[type=submit]').trigger('click');

      setTimeout(() => {
        expect(wrapper.find('.success').exists()).toBeTruthy();
        done();
      });
    });

    it('Should store token when request succeed.', done => {
      when(validateEmail as jest.Mock).calledWith(email).mockReturnValue(true);

      wrapper.setData({ password, email });

      wrapper.find('[type=submit]').trigger('click');

      setTimeout(() => {
        expect(setItemMock).toBeCalledWith(TOKEN_STORAGE_KEY, token);
        done();
      });
    });
  });
});
