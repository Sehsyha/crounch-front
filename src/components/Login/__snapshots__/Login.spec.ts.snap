// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Should render. 1`] = `
<div>
  <h2>Login</h2>
  <form>
    <div id="email-error" class="error">Please enter a valid email.</div>
    <div id="password-error" class="error">Please enter a password.</div>
    <!----> <input type="text" placeholder="Email"> <input type="password" placeholder="Password"> <input type="submit" value="Se connecter">
  </form>
</div>
`;
